crates/ironrdp-pdu/src/clipboard/capabilities.rs

```diff
@@ -0,0 +1,249 @@
+use crate::clipboard::PartialHeader;
+use crate::cursor::{ReadCursor, WriteCursor};
+use crate::{ensure_fixed_part_size, invalid_message_err, PduDecode, PduEncode, PduResult};
+use bitflags::bitflags;
+
+/// Represents `CLIPRDR_CAPS`
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Capabilities {
+    pub capabilities: Vec<CapabilitySet>,
+}
+
+impl Capabilities {
+    const NAME: &str = "CLIPRDR_CAPS";
+    const FIXED_PART_SIZE: usize = std::mem::size_of::<u16>() * 2;
+
+    fn inner_size(&self) -> usize {
+        Self::FIXED_PART_SIZE + self.capabilities.iter().map(|c| c.size()).sum::<usize>()
+    }
+}
+
+impl PduEncode for Capabilities {
+    fn encode(&self, dst: &mut WriteCursor<'_>) -> PduResult<()> {
+        if self.capabilities.len() > u16::MAX as usize {
+            return Err(invalid_message_err!(
+                "cCapabilitiesSets",
+                "Too much capability sets specified",
+            ));
+        }
+
+        let header = PartialHeader::new(self.inner_size() as u32);
```

Comment:
I suggest either using `u32::from` or the `cast_length!` macro if the conversion is fallible. Same for all other `as` conversions because those are lossy conversions while we need lossless conversion (if the value does not fit in the target type, we have a problem). Since this is not exactly what I would call "elements", you could also add a new `cast_int!` macro like this one:
```rust
#[macro_export]
macro_rules! cast_int {
    ($len:expr, $name:expr, $field:expr) => {{
        $len.try_into().map_err(|_| $crate::Error::InvalidMessage {
            name: $name,
            field: $field,
            reason: "out of range integral type conversion",
        })
    }};
}
```

Reply:
Oh, good point, indeed this is an oversight, I'll make `PartialHeader::new` return `Result` and perform usize->u32 cast with validation inside, Macro would be great addition too!

Reply:
Sounds good, but I do think it’s also okay for `PartialHeader::new` to stay the way it currently is:
```rust
let inner_size: u32 = cast_int!(self.inner_size());
let header = PartialHeader::new(inner_size);
```